File: Binary search/154_Find Minimum in Rotated Sorted Array II.cpp
/*
由于数组之前是递增的，所以旋转后也还是总有一个区间是递增的
就不停的找递增区间，他最小值很容易找到，然后再对另一个区间进行二分。
*/
class Solution {
public:
    int findMin(vector<int>& nums) {
        if(nums.size()==1) return nums[0];
        int l=0,r=nums.size()-1;
        int min=nums[0];
        while(l<=r){            
            int mid=(l+r)/2;
            if(nums[mid]==nums[l]){
                if(min>=nums[l]) min=nums[l];
                ++l;
            }
            else if(nums[mid]<=nums[r]){
                if(min>=nums[mid]) min=nums[mid];
                r=mid-1;
            }
            else{
                if(min>nums[l]) min=nums[l];
                l=mid+1;
            }
        }
        return min;
    }
};

File: Binary search/154_Find Minimum in Rotated Sorted Array II.py
#1.与CPP中一致的代码：
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if len(nums)==1:
            return nums[0]
        l,r=0,len(nums)-1
        aim=nums[r]
        while l<=r:
            mid=(r+l)//2
            if nums[mid]==nums[l]:
                if aim>nums[l]:
                    aim=nums[l]
                l+=1
            elif nums[mid] <=nums[r]:
                if aim>nums[mid]:
                    aim = nums[mid]
                r=mid-1
            else:
                if aim>nums[l]:
                    aim=nums[l]
                l=mid+1
        return aim
    
#更简洁的方法：
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
        while left < right:
            mid = (left + right) // 2
            if nums[mid] > nums[right]: left = mid + 1
            elif nums[mid] < nums[right]: right = mid
            else: right = right - 1 # key
        return nums[left]
